apps/andi/lib/andi/input_schemas/ingestion.ex

```diff
@@ -0,0 +1,172 @@
+  alias Andi.InputSchemas.Datasets.ExtractStep
+  alias Andi.Schemas.Validation.CadenceValidator
+  alias AndiWeb.Helpers.ExtractStepHelpers
+  alias AndiWeb.Views.Options
+  alias Andi.InputSchemas.DatasetSchemaValidator
+
+  @primary_key {:id, Ecto.UUID, autogenerate: true}
+  schema "ingestions" do
+    field(:allow_duplicates, :boolean)
+    field(:cadence, :string)
+    field(:sourceFormat, :string)
+    field(:topLevelSelector, :string)
+    belongs_to(:dataset, Dataset, type: :string, foreign_key: :targetDataset)
+    has_many(:schema, DataDictionary, on_replace: :delete)
+    has_many(:extractSteps, ExtractStep, on_replace: :delete)
+  end
+
+  use Accessible
+
+  @cast_fields [
+    :id,
+    :cadence,
+    :sourceFormat,
+    :topLevelSelector,
+    :targetDataset
+  ]
+
+  @required_fields [
+    :cadence,
+    :sourceFormat,
+    :targetDataset
+  ]
+
+  @submission_cast_fields [
+    :sourceFormat
+  ]
+
+  @submission_required_fields [
+    :sourceFormat
+  ]
+
+  def changeset(%SmartCity.Ingestion{} = changes) do
+    changes_as_map = StructTools.to_map(changes)
+    changeset(%__MODULE__{}, changes_as_map)
+  end
+
+  def changeset(changes), do: changeset(%__MODULE__{}, changes)
+
+  def changeset(ingestion, changes) do
+    changes_with_id = StructTools.ensure_id(ingestion, changes)
+    source_format = Map.get(changes, :sourceFormat, nil)
+
+    ingestion
+    |> cast(changes_with_id, @cast_fields, empty_values: [])
+    |> validate_required(@required_fields, message: "is required")
+    |> cast_assoc(:schema, with: &DataDictionary.changeset(&1, &2, source_format), invalid_message: "is required")
+    |> cast_assoc(:extractSteps, with: &ExtractStep.changeset/2)
+    |> foreign_key_constraint(:targetDataset)
+    |> validate_source_format()
+    |> CadenceValidator.validate()
+    |> validate_top_level_selector()
+    |> validate_schema()
+    |> validate_extract_steps()
+  end
+
+  def submission_changeset(ingestion, changes) do
+    changes_with_id = StructTools.ensure_id(ingestion, changes)
+    source_format = Map.get(changes, :sourceFormat, nil)
+
+    ingestion
+    |> cast(changes_with_id, @submission_cast_fields, empty_values: [])
+    |> cast_assoc(:schema, with: &DataDictionary.changeset(&1, &2, source_format), invalid_message: "is required")
+    |> foreign_key_constraint(:targetDataset)
+    |> validate_required(@submission_required_fields, message: "is required")
+    |> validate_source_format()
+    |> validate_submission_schema()
+  end
+
+  def changeset_for_draft(ingestion, changes) do
+    changes_with_id = StructTools.ensure_id(ingestion, changes)
+
+    ingestion
+    |> cast(changes_with_id, @cast_fields, empty_values: [])
+    |> cast_assoc(:schema, with: &DataDictionary.changeset_for_draft/2)
+    |> cast_assoc(:extractSteps, with: &ExtractStep.changeset_for_draft/2)
+    |> foreign_key_constraint(:targetDataset)
+  end
+
+  def preload(struct), do: StructTools.preload(struct, [:schema, :extractSteps])
```

Comment:
Should "transformations" be here as well

Edit: On second thought, it looks like transformations is absent from other parts of this PR

I'm assuming now that we'll probably add transformations logic to ingestions once we're adding those for the first time. But for now, this PR restores the functionality of adding / removing datasets.

Disregard this comment then

Reply:
Thanks, Ben! Exactly, transformations was out of scope for the initial ingestions work.